Added SAPM Exporter

owais opened the pull request:
SAPM exporter implements a trace exporter for the SAPM protocol.

The exporter is implemented as a thin wrapper over the SAPM protocol
HTTP client (https://github.com/signalfx/sapm-proto/tree/master/client)

Depends on sapm-proto 0.1.0 being released after merging: https://github.com/signalfx/sapm-proto/pull/8

owais commented on go.mod:
```diff
@@ -9,15 +9,16 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/collectdreceiver v0.0.0-20191209163404-28d5712f4129
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/sapmreceiver v0.0.0-20191209183045-9707a2732f37
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/zipkinscribereceiver v0.0.0-20191126142441-b2a048090ad6
-	github.com/open-telemetry/opentelemetry-collector-contrib/testbed v0.0.0-20191209183045-9707a2732f37 // indirect
```
Removed by `go mod tidy`. Probably wasn't needed?

pjanotti commented on exporter/sapmexporter/config.go:
```diff
@@ -0,0 +1,29 @@
+type Config struct {
+	configmodels.ExporterSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct.
+	Endpoint                      string                   `mapstructure:"endpoint"`
+	AuthToken                     string                   `mapstructure:"auth_token"`
```
nit: per comment in the signalfx exporter PR, use `access_token` instead of `auth_token`, see https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38#discussion_r348496782

pjanotti replied:
nit: add comments to each field here.

pjanotti commented on exporter/sapmexporter/exporter.go:
```diff
@@ -0,0 +1,101 @@
+	logger *zap.Logger
+}
+
+func (*sapmExporter) Name() string {
```
I don't think this is needed.

pjanotti replied:
https://github.com/open-telemetry/opentelemetry-collector/issues/456

pjanotti commented on go.mod:
```diff
@@ -9,15 +9,16 @@ require (
 	honnef.co/go/tools v0.0.1-2019.2.3
 )
+
+replace github.com/open-telemetry/opentelemetry-collector-contrib/exporter/sapmexporter => ./exporter/sapmexporter
```
nit: can we have all replaces here on the bottom? there is a previous one on the top

pjanotti commented on exporter/sapmexporter/testdata/config.yaml:
```diff
@@ -0,0 +1,24 @@
+exporters:
+  sapm:
+  sapm/customname:
+    endpoint: test-endpoint
```
nit: comment each option used here, so we can point users to it as a sample

pjanotti commented on exporter/sapmexporter/exporter.go:
```diff
@@ -0,0 +1,101 @@
+}
+
+func (se *sapmExporter) Shutdown() error {
+	se.client.Stop()
```
Q.: What happens with any flying request? Does it wait for them to finish?

owais replied:
Yes, it waits for them to finish and if more requests come in while it is waiting then it waits for them as well. In practice it can accept <N more requests after Stop() is called and finish them before Stop() returns where N is number of workers.

pjanotti reviewed (commented):
LGTM a few nits and one Q.

pjanotti commented on exporter/sapmexporter/exporter.go:
```diff
@@ -0,0 +1,101 @@
+	if err != nil {
+		return 0, consumererror.Permanent(err)
+	}
+	err = se.client.Export(ctx, jBatch)
```
@owais what this client does that it has workers/etc? In general is better to have something that simply sends the data (no batching or retrying on its own).

pjanotti replied:
@owais as you know in general is better to avoid batching and retries on the exporter, not a deal breaker but not recommended in general. Any plans to make it a simple sender down the road?

tigrannajaryan reviewed (approved):
Generally LGTM, please address @pjanotti's comments.

owais committed:
Added SAPM Exporter

SAPM exporter implements a trace exporter for the SAPM protocol.

The exporter is implemented as a thin wrapper over the SAPM protocol
HTTP client (1).

1: https://github.com/signalfx/sapm-proto/tree/master/client

pjanotti reviewed (approved):
LGTM
